drivers/hmis/lib/hmis_util/json_forms.rb

```diff
@@ -81,18 +100,65 @@ def walk_nodes(node, &block)
       children&.each { |child| walk_nodes(child, &block) }
     end
 
-    def apply_fragment(item)
-      return item unless item['fragment']
+    def resolve_fragments!(base_item)
+      walk_nodes(base_item) do |item|
+        next unless item['fragment'].present?
+
+        fragment_key = item['fragment']&.gsub(/^#/, '')
+        fragment = fragment_map[fragment_key]
+        raise "Fragment not found #{item['fragment']}" unless fragment.present?
+
+        fragment_items = fragment['item'] || [] # child items of the fragment
+        additional_items = item['item'] || [] # any items that should be appended
+
+        # Reverse merge so that any keys specified in 'item' overried the fragment values.
+        # This can be useful in changing the link id, text, etc.
+        # This is a shallow merge.
+        item.reverse_merge!(fragment)
+        # If this item was adding any additional items, we need to add the fragment items
+        # since they wouldn't have been copied by the shallow merge
+        item['item'].unshift(*fragment_items) if additional_items.any? && fragment_items.any?
+        # Remove the fragment field
+        item.delete('fragment')
+      end
+    end
+
+    def resolve_all_fragments!(definition)
+      definition['item'].each { |i| resolve_fragments!(i) }
+    end
+
+    # This function creates/updates a FormDefinition, and applies any fragments and patches.
+    #
+    # Fragments are for re-using questions across different assessments.
+    # Patches are for applying installation-specific changes to those patches.
+    #
+    # First, we "resolve" any patches that are referenced throughout the form.
+    # Next, we apply any installation-specific patches, which could make
+    # any arbitrary changes to any item (by Link ID)
+    #
+    # This approach lets you do something like change the link id for a fragment (at resolution time)
+    # and then apply a patch just to that version. A use-case would be if you
+    # want to change something about Disability fragment just for Intake,
+    # not other assessments.
```

Comment:
This here is a bit different from how you implemented it @ttoomey. I ended up needing a bit more flexibility than I thought.
It's not perfect, still, though. With this approach we need to be careful with uniqeness of link ids, since the patches get applied to ANY definition that gets loaded. So if we have Link IDs in multiple forms that are the same, and there is a patch referencing it, it will get applied in both places. (Which is what we want often times, but I can also see it happening accidentally).